src/uct/base/uct_mem.c

```diff
@@ -99,15 +99,21 @@ ucs_status_t uct_mem_alloc(void *addr, size_t min_length, unsigned flags,
                     return status;
                 }
 
-                /* Check if MD supports allocation */
-                if (!(md_attr.cap.flags & UCT_MD_FLAG_ALLOC)) {
-                    continue;
-                }
-
-                /* Check if MD supports allocation with fixed address
-                 * if it's requested */
-                if ((flags & UCT_MD_MEM_FLAG_FIXED) &&
-                    !(md_attr.cap.flags & UCT_MD_FLAG_FIXED)) {
+                if (!(flags & UCT_MD_MEM_FLAG_ON_DEVICE)) {
+                    /* Check if MD supports allocation */
+                    if (!(md_attr.cap.flags & UCT_MD_FLAG_ALLOC)) {
+                        continue;
+                    }
+
+                    /* Check if MD supports allocation with fixed address
+                     * if it's requested */
+                    if ((flags & UCT_MD_MEM_FLAG_FIXED) &&
+                        !(md_attr.cap.flags & UCT_MD_FLAG_FIXED)) {
+                        continue;
+                    }
+                } else if (!(md_attr.cap.flags & UCT_MD_FLAG_ALLOC_DEVICE_MEM)) {
```

Comment:
To every "continue" - add code comment why we skip the MD

- before any loop - check that FIXED is not with DEVICE_MEM - return INVALID_PARAM
- if user asked ALLOC_DEVICE, MD does not have it -> continue
- if user did NOT ask ALLOC_DEVICE, MD does not have ALLOC -> continue
- if user asked FIXED, MD does not have fixed -> continue (as before)